File: 0001-two-sum/0001-two-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    //brute force soln:time complexity is o(n^2)
    // for(let i=0;i<nums.length;i++) {
    //     for(let j=i+1;j<nums.length;j++) {
    //         const sum = nums[i] + nums[j];
    //         const index = [i,j];
    //         if(sum===target) return index; 
    //     }
    // }

    const map = new Map();

    for(let i=0;i<nums.length;i++) {
        //calculate the complement
        const complement = target - nums[i];
        if(map.has(complement)) return [map.get(complement),i] //map.get(complement) will return the complemen's index, i-> current index
        //store the number in the map {number,index}
        map.set(nums[i],i);
    }
};

File: 0001-two-sum/0001-two-sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        HashMap <Integer,Integer> map = new HashMap<>();
        int complement;
        //using complement to solve the question hash map will store the num:index
        for(int i=0;i<nums.length;i++) {
            complement = target - nums[i];
            if(map.containsKey(complement)) {
                return new int[] {map.get(complement), i};
            }

            //store the values in map
            map.put(nums[i], i);
        }

        return new int[] {-1,-1};
    }
}